### ISSUE TITLE
Operations for collaborative editing

### ISSUE THREAD

**Issue by streamich** [excalidraw/excalidraw#3210]
Does Excalidraw emit "operations" on every user action, so one can subscribe to those operations and use the operations to send to the server to support collaborative editing?

**Comment by dwelle** [comment]
Not at the moment. We currently only support a `onChange` callback which gets invoked on every change with the full scene data `{elements, appState}`, where you'll have to do the diffing yourself if you need deltas (but the data is also mutable so that won't help).

At excalidraw, we only sync changed element(s) by keeping around a map of last-emitted elements (by checking `version`). You can check the code here: https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/CollabWrapper.tsx#L565 and https://github.com/excalidraw/excalidraw/blob/master/src/excalidraw-app/collab/Portal.tsx#L87

**Comment by dwelle** [comment]
Yes. We expose several APIs from the Excalidraw instance, which you can reach by passing a `ref` to the component. In particular, you'll want to use the `updateScene()` API. Check the readme for an example:

https://github.com/excalidraw/excalidraw/blob/master/src/packages/excalidraw/README.md#usage

**Comment by streamich** [comment]
Thanks, I was looking for it on the `ref` but could not find it, I was looking for something like `updateState`, did not occur to me that it is `updateScene` 😁 

I think this should be enough for what I'm doing. I'm trying to connect Excalidraw to the new [Multiplayer](https://finalfloat.com/products/multiplayer) tool.

**Comment by dwelle** [comment]
Cool. Be sure to drop us a note when it's ready!

**Comment by streamich** [comment]
If you click "Excalidraw sketchpad" in the footer you can play with the demo, [here](https://finalfloat.com/):

[![image](https://user-images.githubusercontent.com/9773803/110713464-4723b500-8202-11eb-9a56-d9da9a77710f.png)](https://finalfloat.com/)

It is tricky to implement collab editing using only the public `.updateScene()` and `onChange` APIs, both of which operate on the whole state object. But it is kind of possible to compute state differences and send only those differences over the network. It partially works, is still a bit buggy, but some good progress for a simple demo and half-day of experimentation.

**Comment by streamich** [comment]
Would be nice for Excalidraw to expose proper public APIs that allow it to be used as a controlled component: it always renders just the state it is given and a callback receives back changes that Excalidraw wants to do on the state object, and then it is up to the user to apply those changes to the sate object. And once the state changes are applied it would re-render without losing focus of the active element. Otherwise now (without forking Excalidrwa) the changes need to be computed manually and I believe the focus is lost when the state is updated when new data is received from the server.

**Comment by dwelle** [comment]
I think making Excalidraw into a controlled component would kill performance as there can be tens (or more) changes in a second. Something that a top-down flow would make much slower (but we'd have to benchmark).

Unclear what you mean by losing focus on `updateScene()` call. We're using this API on excalidraw.com ourselves (when collaborating).
